### PR TITLE
Use shaded dependency on JCTools instead of copy and paste

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:
JCTools supports both non-unsafe, unsafe versions of queues and JDK6 which allows us to shade the library in netty-common allowing it to stay "zero dependency".

Modifications:
- Remove copy paste JCTools code and shade the library (dependencies that are shaded should be removed from the <dependencies> section of the generated POM).
- Remove usage of OneTimeTask

Result:
Less code to maintain and easier to update JCTools and less GC pressure as the queue implementation nt creates so much garbage

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin @nmittler @Scottmitch @guidomedina @nitsanw PTAL... hoorrrayyy

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java [hunk 1]
```diff
@@ -273,7 +272,7 @@ public final void finishHandshake(Channel channel, FullHttpResponse response) {
             // Delay the removal of the decoder so the user can setup the pipeline if needed to handle
             // WebSocketFrame messages.
             // See https://github.com/netty/netty/issues/4533
-            channel.eventLoop().execute(new OneTimeTask() {
+            channel.eventLoop().execute(new Runnable() {
```
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: Is the change to `Runnable` related to JCTools? Could you factor this out into a separate PR?

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: @nmittler I could but Im not sure we should as basically the hack that we had with OneTimeTask is not needed anymore because we not use a queue that is based on a linked node structure. So this is related and so I think it should stay here.

File: common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java [hunk 2]
```diff
@@ -18,6 +18,10 @@
 
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 
+/**
+ * @deprecated will be removed as not used anymore inside netty.
```
Comment (nmittler) [inline common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java]: Maybe expand the comment to include the JCTools replacement class?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java]: There is really no replacement ;) As this is not needed anymore

**Comment by nmittler** [comment]
@normanmaurer this is awesome! :)

We should run a full suite of benchmarks before and after to make sure there are no performance regressions.

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 3]
```diff
@@ -1009,13 +1007,15 @@ public String toString() {
...
+        private final Recycler.Handle<AbstractWriteTask> handle;
         private AbstractChannelHandlerContext ctx;
         private Object msg;
         private ChannelPromise promise;
         private int size;
 
+        @SuppressWarnings("unchecked")
         private AbstractWriteTask(Recycler.Handle<? extends AbstractWriteTask> handle) {
-            super(handle);
+            this.handle = (Recycler.Handle<AbstractWriteTask>) handle;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: MINOR Remove this unnecessary cast to "Handle". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1905)

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 4]
```diff
@@ -607,30 +613,24 @@ public static int hashCodeAscii(CharSequence bytes) {
      * consumer (one thread!).
      */
     public static <T> Queue<T> newMpscQueue() {
-        return new MpscLinkedQueue<T>();
+        return hasUnsafe() ? new MpscChunkedArrayQueue<T>(1024, MAX_MPSC_CAPACITY, true)
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: MAJOR Ternary operators that can be simplified with || or && [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ASimplifiedTernary)

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 5]
```diff
@@ -82,6 +87,7 @@
     private static final boolean DIRECT_BUFFER_PREFERRED =
             HAS_UNSAFE && !SystemPropertyUtil.getBoolean("io.netty.noPreferDirect", false);
     private static final long MAX_DIRECT_MEMORY = maxDirectMemory0();
+    private static final int MAX_MPSC_CAPACITY =  1024 * 1024; // TODO: Maybe make this configurable ?
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: INFO Complete the task associated to this TODO comment. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

Comment (guidomedina) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: netkins is bossy eh.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: lol... still free to ignore it tho...

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java (hunk 3 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: this is needed

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java (hunk 4 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: this is fine

File: common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java (hunk 2 above)
Comment (johnou) [inline common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java]: Then why not just remove it? It is in the internal package after all.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java]: Because it is used by OneTimeTask and I know projects use OneTimeTask. The idea was to make it as smooth as possible to upgrade.

**Comment by guidomedina** [comment]
@normanmaurer because I took some risks on my own and built my own pre-view of this PR for my project I went further and optimized `PlatformDependent` even more, you might want to include these changes, which you can see at https://github.com/guidomedina/netty/commit/762ac8a6c5e412ba80fe5a28ef189a21b4155e97

The only thing different from what you have is the fact that I replaced all calls inside `PlatformDependent` from functions to constants, the rest is the same.

**Comment by guidomedina** [comment]
Note: About previous; it doesn't have the `OneTimerTask` changes so you will have to kind of copy & paste my `PlatformDependent` and diff if you would like these changes, or cherry pick and rebase your commit, whichever (if you think is worth) is easier for you.

**Comment by normanmaurer** [comment]
@guidomedina thanks but actually it should not make any difference as these will be inlined anyway

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java (hunk 1 above)
Comment (johnou) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: Would you mind explaining why the hack was required in the first place?

**Comment by guidomedina** [comment]
@normanmaurer will it be possible to port this PR to 4.0.x as you deprecated some classes it wouldn't harm projects using such classes right?

**Comment by guidomedina** [comment]
I'm using Vert.x 3.x websockets to stream many small messages which is based on top of Netty 4.0.x for the moment, I can see it benefiting a lot from this PR for example, I tested it Vert.x 3.x with Netty 4.1.0.Final and it is not fully compatible, no errors but is not working, it might take sometime to migrate.

**Comment by normanmaurer** [comment]
Yes thats the plan

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java (hunk 1 above)
Comment (buchgr) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: @johnou just a hack to save on object allocation. The `OneTimeTask` is both a `Runnable` and a `LinkedQueueNode`, so you don't need an extra `LinkedQueueNode` when adding it to the queue.

File: common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java (hunk 2 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/MpscLinkedQueueNode.java]: This class is internal. My vote is to remove it if we no longer need it. Lean and mean code base.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: I think it makes sense to remove `OneTimeTask` in this PR as its existence dependents upon `MpscLinkedQueueNode` which should be removed in this PR.

File: common/src/main/java/io/netty/util/internal/OneTimeTask.java [hunk 6]
```diff
@@ -15,14 +15,11 @@
...
 
 /**
- * {@link Runnable} which represent a one time task which may allow the {@link EventExecutor} to reduce the amount of
- * produced garbage when queue it for execution.
- *
- * <strong>It is important this will not be reused. After submitted it is not allowed to get submitted again!</strong>
+ * @deprecated Use {@link Runnable}.
  */
+@Deprecated
 public abstract class OneTimeTask extends MpscLinkedQueueNode<Runnable> implements Runnable {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/OneTimeTask.java]: This class is in an internal package. My vote is to remove it if we no longer need it. Lets try to keep the code base lean (for Netty :wink:) and mean where possible. Also using `OneTimeTask` will no longer provide the memory savings which is was most likely used for.

File: common/src/main/java/io/netty/util/HashedWheelTimer.java [hunk 7]
```diff
@@ -567,11 +565,6 @@ public boolean isExpired() {
             return state() == ST_EXPIRED;
         }
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/HashedWheelTimer.java]: line 537 there is a comment which references MpscLinkedQueueNode ... plz remove.

**Comment by Scottmitch** [comment]
@normanmaurer - is there a way to declare the `maven-shade-plugin` as compile scope so it doesn't get put into the deployed pom?

**Comment by normanmaurer** [comment]
@Scottmitch no I don't think there is a way... What is the problem you have with have it included ?

File: common/src/main/java/io/netty/util/internal/OneTimeTask.java (hunk 6 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/OneTimeTask.java]: @Scottmitch well Im not super strong here but I think it also is not really hurting to keep it for one or two more releases... There is no "overhead" in just keep it. @trustin WDYT ?

Comment (trustin) [inline common/src/main/java/io/netty/util/internal/OneTimeTask.java]: I'm fine with either options. It's an internal class, so I think it's fine to remove right away, but if @normanmaurer wishes to take care of removal later, then I'm also fine with that.

File: common/src/main/java/io/netty/util/HashedWheelTimer.java (hunk 7 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/HashedWheelTimer.java]: will do... Also I can simplify this code a bit now that we not use MpscLinkedQueueNode anymore :)

File: common/src/main/java/io/netty/util/internal/OneTimeTask.java (hunk 6 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/OneTimeTask.java]: @nmittler WDYT ;) ?

File: common/src/main/java/io/netty/util/HashedWheelTimer.java [hunk 8]
```diff
@@ -530,25 +528,21 @@ public boolean cancel() {
             if (!compareAndSetState(ST_INIT, ST_CANCELLED)) {
                 return false;
             }
-            // If a task should be canceled we create a new Runnable for this to another queue which will
+            // If a task should be canceled we put his to another queue which will
```
Comment (johnou) [inline common/src/main/java/io/netty/util/HashedWheelTimer.java]: put this?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/HashedWheelTimer.java]: yes ... thanks!

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java (hunk 1 above)
Comment (johnou) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: @buchgr makes sense :+1:

**Comment by Scottmitch** [comment]
> no I don't think there is a way... What is the problem you have with have it included ?

@normanmaurer - No problem. Just noticed that it is there in the published artifact and I'm not sure that it needs to be.

File: common/src/main/java/io/netty/util/internal/OneTimeTask.java (hunk 6 above)
Comment (nmittler) [inline common/src/main/java/io/netty/util/internal/OneTimeTask.java]: @normanmaurer I'm leaning toward just killing it ... not super strong either way though.

**Comment by normanmaurer** [comment]
4.1.1.Final and 4.0.37.Final are due this week. Will merge this once these are out.

**Comment by normanmaurer** [comment]
@nmittler @buchgr can you guys have a full benchmark suite run for grpc with this one and without ?

**Comment by nmittler** [comment]
@buchgr could you run some grpc benchmarks on this?

**Comment by buchgr** [comment]
@nmittler sure!

**Comment by buchgr** [comment]
@nmittler @normanmaurer ran some quick grpc benchmarks. I took the best out of three runs. Numbers seem identical.

**4.1.1.Final**
```
./qps_client --address=localhost:8080 --directexecutor
Channels:                       4
Outstanding RPCs per Channel:   10
Server Payload Size:            0
Client Payload Size:            0
50%ile Latency (in micros):     423
90%ile Latency (in micros):     459
95%ile Latency (in micros):     480
99%ile Latency (in micros):     656
99.9%ile Latency (in micros):   1407
Maximum Latency (in micros):    19215
QPS:                            91860
```
```
Benchmark                         (direct)  (transport)    Mode    Cnt       Score     Error  Units
TransportBenchmark.unaryCall1024      true        NETTY  sample  99610  100234.723 ± 562.832  ns/op
TransportBenchmark.unaryCall1024     false        NETTY  sample  91764  108805.391 ± 601.117  ns/op
```

**This PR**
```
Jakobs-MBP-4:bin buchgr$ ./qps_client --address=localhost:8080 --directexecutor
Channels:                       4
Outstanding RPCs per Channel:   10
Server Payload Size:            0
Client Payload Size:            0
50%ile Latency (in micros):     426
90%ile Latency (in micros):     463
95%ile Latency (in micros):     489
99%ile Latency (in micros):     701
99.9%ile Latency (in micros):   1371
Maximum Latency (in micros):    9719
QPS:                            91313
```
```
Benchmark                         (direct)  (transport)    Mode     Cnt       Score     Error  Units
TransportBenchmark.unaryCall1024      true        NETTY  sample  101283   98596.506 ± 563.207  ns/op
TransportBenchmark.unaryCall1024     false        NETTY  sample   92437  108009.572 ± 632.538  ns/op
```

**Comment by normanmaurer** [comment]
@buchgr actually can you use "NOT" the direct executor ? I would like to have stuff triggered from outside the EventLoop if possible 👍

**Comment by normanmaurer** [comment]
Ah sorry... seems like you benchmarked both... Then looks good to me for merge... @Scottmitch @nmittler ok ?

**Comment by Scottmitch** [comment]
@normanmaurer - lgtm! yaay for no more copy/paste!

**Comment by buchgr** [comment]
@normanmaurer QPS tests without direct executor. seems fine.

**4.1.1.Final**
```
./qps_client --address=localhost:8080
Channels:                       4
Outstanding RPCs per Channel:   10
Server Payload Size:            0
Client Payload Size:            0
50%ile Latency (in micros):     720
90%ile Latency (in micros):     2093
95%ile Latency (in micros):     4179
99%ile Latency (in micros):     13103
99.9%ile Latency (in micros):   46175
Maximum Latency (in micros):    136191
QPS:                            28731
```

**This PR**
```
./qps_client --address=localhost:8080
Channels:                       4
Outstanding RPCs per Channel:   10
Server Payload Size:            0
Client Payload Size:            0
50%ile Latency (in micros):     704
90%ile Latency (in micros):     2099
95%ile Latency (in micros):     4195
99%ile Latency (in micros):     13303
99.9%ile Latency (in micros):   44127
Maximum Latency (in micros):    356095
QPS:                            29026
```

**Comment by Scottmitch** [comment]
@buchgr - Looks like maximum latency increased ... is this an anomaly?

**Comment by buchgr** [comment]
@Scottmitch yep. It's mostly random and not of any meaning really. I see fluctuations by a factor of ten between runs. We should probably not report it. The benchmark did 29026 * 60s = 1.7 million RPC calls. The latency of one out of 1.7 million should not be significant. The percentiles are much more meaningful and those are stable between runs.

**Comment by nmittler** [comment]
@normanmaurer ship it!

**Comment by Scottmitch** [comment]
@buchgr - thanks for confirming and running the benchmarks!

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (c784271620cb94a4dce99057a257ada58ffdb829) and 4.1 (c3abb9146e47b1c8de5a26b7ae5076845fc80c28)

**Comment by normanmaurer** [comment]
Thanks again to all of you!

**Comment by normanmaurer** [comment]
ups... done... Thanks @slandelle !

**Comment by nitsanw** [comment]
Huge thanks to @normanmaurer @guidomedina and everyone else who helped push this change :-)
